### PR TITLE
Alerting: show state history

### LINKED ISSUE #41164: Unified Alerting: Display 'Alert History' via annotations (same as previous alerting)
Currently, in both legacy and Grafana v8 alerts. When an alert changes state, an annotation is created with the transition.

In the legacy alerting you could see these state changes as some sort of "alert history".
Screenshot 2021-11-01 at 17 38 02
Screenshot 2021-11-01 at 17 38 34

The annotations API endpoint supports per rule loading, so technically, it is possible to do an API call that filters all annotations for a given alert rule.

Can we support this in some form in the front-end to ensure there's backward compatibility for this feature?

### PR DESCRIPTION
**Pull request by gillesdemey**
**What this PR does / why we need it**:

Adds a simple modal to show the alert history for Grafana managed alerts.

image

**Which issue(s) this PR fixes**:

Fixes #41164

**Special notes for your reviewer**:

~Currently in draft, needs some tests and a validation that this does not break anything for Prometheus / Cortex / Loki alert rules~

The alert state history is currently only available for Grafana managed alerts, cloud alerts do not have their state changes recorded in the `/api/annotations` API.

### CODE REVIEW

File: public/app/features/alerting/unified/utils/rules.ts [hunk 1]
```diff
@@ -71,6 +72,34 @@ export function alertStateToReadable(state: PromAlertingRuleState | GrafanaAlert
   return capitalize(state);
 }
 
+export function LegacyAlertStateToGrafanaAlertState(legacyState: AlertState): GrafanaAlertState {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/rules.ts:75]: Not too sure about this function; I did not find anything similar but the annotations still use the older legacy `AlertState` states.

File: public/app/features/alerting/unified/components/rules/RuleStateHistory.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+const RuleStateHistory: FC<RuleStateHistoryProps> = ({ alertId }) => {
+  const { loading, error, result = [] } = useManagedAlertStateHistory(alertId);
+
+  if (loading && !error) {
+    return <LoadingPlaceholder text={'Loading history...'} />;
+  }
+
+  if (error && !loading) {
+    return <div>{error.toString()}</div>;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/RuleStateHistory.tsx:27]: This one needs a better error display, any pointers?

File: public/app/features/alerting/unified/components/rules/RuleDetailsActionButtons.tsx [hunk 3]
```diff
@@ -115,7 +118,7 @@ export const RuleDetailsActionButtons: FC<Props> = ({ rule, rulesSource }) => {
           <LinkButton
             className={style.button}
             size="xs"
-            key="dashboard"
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/RuleDetailsActionButtons.tsx:118]: This one was a duplicate of another button, so made it unique

File: public/app/features/alerting/unified/components/rules/RuleDetailsActionButtons.tsx [hunk 4]
```diff
@@ -172,6 +186,7 @@ export const RuleDetailsActionButtons: FC<Props> = ({ rule, rulesSource }) => {
     if (isViewMode) {
       rightButtons.push(
         <ClipboardButton
+          key="copy"
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/RuleDetailsActionButtons.tsx:189]: This one was missing a `key` and the dev console was complaining about it :)

File: public/app/features/alerting/unified/components/rules/RuleStateHistory.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      label: '',
+      renderCell: renderValueCell,
+    },
+    { id: 'timestamp', label: 'Time', renderCell: renderTimestampCell },
+  ];
+
+  const items: Array<DynamicTableItemProps<PartialHistoryItem>> = result.map((historyItem) => ({
+    id: historyItem.id,
+    data: {
+      newState: historyItem.newState,
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/RuleStateHistory.tsx:47]: Maybe I should just pass the entire `historyItem` and not bother with `PartialHistoryItem`? Thoughts?

**Review by peterholmberg — APPROVED** [review]
Super nice! Only nit is to be more consistent with naming. We have RuleStateHistory, AlertHistory and AlertStateHistory.

**Comment by gotjosh** [comment]
I added a backport - I think this is relatively low risk and backend change was already backported.

**Pull request merged**